driver/bpf/fillers.h

```diff
@@ -649,30 +649,162 @@ static __always_inline int bpf_parse_readv_writev_bufs(struct filler_data *data,
+						       const void __user *iovsrc,
+						       unsigned long iovcnt,
+						       long retval,
+						       int flags,
+						       unsigned long *size)
+{
+	const struct compat_iovec *compat_iov;
+	int res = PPM_SUCCESS;
+	unsigned int copylen;
+	int j;
+
+	copylen = iovcnt * sizeof(struct compat_iovec);
+	compat_iov = (const struct compat_iovec *)data->tmp_scratch;
+
+	if (copylen > SCRATCH_SIZE_MAX)
+	{
+		return PPM_FAILURE_FRAME_SCRATCH_MAP_FULL;
+	}
+
+#ifdef BPF_FORBIDS_ZERO_ACCESS
+		if (copylen)
+			if (bpf_probe_read_user((void *)compat_iov,
+						((copylen - 1) & SCRATCH_SIZE_MAX) + 1,
+						(void *)iovsrc))
 #else
-				if (bpf_probe_read_user(&data->buf[off_bounded],
-							to_read & SCRATCH_SIZE_HALF,
-							iov[j].iov_base))
+		if (bpf_probe_read_user((void *)compat_iov,
+					copylen & SCRATCH_SIZE_MAX,
+					(void *)iovsrc))
 #endif
-					return PPM_FAILURE_INVALID_USER_MEMORY;
+			return PPM_FAILURE_INVALID_USER_MEMORY;
+
+	#pragma unroll
+	for (j = 0; j < MAX_IOVCNT_COMPAT; ++j) {
+		if (j == iovcnt)
+			break;
+		// BPF seems to require a hard limit to avoid overflows
+		if (*size == LONG_MAX)
+			break;
+
+		*size += compat_iov[j].iov_len;
+	}
+
+	if ((flags & PRB_FLAG_IS_WRITE) == 0)
+		if (*size > retval)
+			*size = retval;
+
+	if (flags & PRB_FLAG_PUSH_SIZE && res == PPM_SUCCESS) {
+		res = bpf_push_u32_to_ring(data, (uint32_t)*size);
+		CHECK_RES(res);
+	}
+
+	if (flags & PRB_FLAG_PUSH_DATA) {
+		if (*size > 0) {
+			unsigned long off = _READ(data->state->tail_ctx.curoff);
+			unsigned long remaining = *size;
+			int j;
+
+			// The 14 iovec count limit is due to old kernels verifiers
+			// complaining.
+			#pragma unroll
+			for (j = 0; j < MAX_IOVCNT_COMPAT; ++j) {
+				volatile unsigned int to_read;
+
+				if (j == iovcnt)
+					break;
+
+				unsigned long off_bounded = off & SCRATCH_SIZE_HALF;
+				if (off > SCRATCH_SIZE_HALF)
+					break;
+
+				if (compat_iov[j].iov_len <= remaining)
+					to_read = compat_iov[j].iov_len;
+				else
+					to_read = remaining;
+
+				if (to_read > SCRATCH_SIZE_HALF)
+					to_read = SCRATCH_SIZE_HALF;
+
+	#ifdef BPF_FORBIDS_ZERO_ACCESS
+					if (to_read)
+						if (bpf_probe_read_user(&data->buf[off_bounded],
+									((to_read - 1) & SCRATCH_SIZE_HALF) + 1,
+									(void*)compat_iov[j].iov_base))
+	#else
+					if (bpf_probe_read_user(&data->buf[off_bounded],
+								to_read & SCRATCH_SIZE_HALF,
+								(void*)compat_iov[j].iov_base))
+	#endif
+						return PPM_FAILURE_INVALID_USER_MEMORY;
 
 				remaining -= to_read;
 				off += to_read;
 			}
 		} else {
-			size = 0;
+			*size = 0;
 		}
 
+		return PPM_SUCCESS;
+	}
+	return res;
+}
+#endif
+
+static __always_inline int bpf_parse_readv_writev_bufs(struct filler_data *data,
+						       const void __user *iovsrc,
+						       unsigned long iovcnt,
+						       long retval,
+						       int flags)
+{
+	unsigned long size = 0;
+	int res = PPM_SUCCESS;
+#if defined(CONFIG_X86_64)
```

Comment:
We can remove this check, it should be safe to call `bpf_in_ia32_syscall` on non x86_64.
It should never be true at this level because `sys_enter/sys_exit` do not support ia32 syscalls on non-x86_64 archs.